File: c++/3.cpp
#include<iostream>
#include<vector>
#include<string>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    struct ListNode{
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
    	if (list1 == NULL) return list2;
        if (list2 == NULL) return list1;
        ListNode *ans = new ListNode(), *res;
        res = ans;
        while(list1 != NULL && list2 != NULL){
        	if (list1->val > list2->val){
                ans->next = list2;
                list2 = list2->next;
            }
        	else{
                ans->next = list1;
                list1 = list1->next;
            }
            ans = ans->next;
        }
        if (list1 == NULL)ans->next = list2;
        else ans->next = list1;
        return res->next;
    }
    vector<string> generateParenthesis(int n) {
    	vector<string> ans;
        return helper(ans, "(", "(", n-1, 2*n);
    }
    vector<string> helper(vector<string> &ans, string real, string res, int n, int len){
        if (res.size() == len){
            ans.push_back(res);
        }
    	if (n){
            helper(ans, real+"(", res+"(", n-1, len);
        }
        if (real.size()){
            helper(ans, real.substr(0,real.size()-1), res+")", n, len);
        }
        return ans;

    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* res = new ListNode();
        ListNode* head = res;
        while(!lists.empty()){
            int min=10001,index = 0;
        	for(int i=0;i < lists.size(); i++){
            	if(lists[i] == NULL){
                	lists.erase(lists.begin() + i);
               	 	i--;
                    continue;
            	}
                if(min > lists[i]->val){
                    min =lists[i]->val;
                    index = i;
                }
            }

            res->next = lists[index];
            res = res->next;
            if (lists[index] != NULL)
            lists[index] = lists[index]->next;
    	}
        return head->next;
    }
    ListNode* swapPairs(ListNode* head) {
        ListNode* pre = head;
        while(pre != NULL){
            if(pre->next != NULL){
                // swap
                int t = pre->val;
                pre->val = pre->next->val;
                pre->next->val = t;
                pre = pre->next->next;
            }else break;
        }
    	return head;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *node = head;
        int cnt = k;
        while (cnt--){
            if (node == NULL)return head;
            node = node->next;
        }
        ListNode *newHead = reverse(head, node);
        head->next = reverseKGroup(node, k);
        return newHead;
    }

    ListNode* reverse(ListNode *first, ListNode *last){
        ListNode *pre = last;
        while(first != last){
            ListNode *tmp = first->next;
            first->next = pre;
            pre = first;
            first = tmp;
        }
        return pre;
    }

    int removeDuplicates(vector<int>& nums) {
        int len = nums.size(), res = 0;
        for(int i=1;i < len-res;i++){
            if(nums[i] == nums[i-1]){
                res++;
                nums.erase(nums.begin()+i);
                nums.push_back(nums[--i]);
            }
    	}
        return len-res;
    }
    int removeElement(vector<int>& nums, int val) {
        int res = 0;
        for(int i=0; i < nums.size(); i++){
            if(nums[i] == val){
                nums.erase(nums.begin()+i);
                i--;
                res++;
            }
        }
        for(int i=0; i < res ; i++){
            nums.push_back(val);
        }
        return nums.size() - res;
    }

    int strStr1(string haystack, string needle) {
        int m = haystack.size(), n = needle.size();
        for (int i = 0; i <= m - n; i++) {
            int j = 0;
            for (; j < n; j++) {
                if (haystack[i + j] != needle[j]) {
                    break;
                }
            }
            if (j == n) {
                return i;
            }
        }
        return -1;
    }
    int strStr2(string haystack, string needle) {
        int m = haystack.size(), n = needle.size(), p = 0;
        while (p + n - 1 < m) {
            if (haystack.substr(p, n) == needle) {
                return p;
            }
            while (p++ + n - 1 < m && haystack[p] != needle[0]);
        }
        return -1;
    }
    int strStr3(string haystack, string needle) {
        int m = haystack.size(), n = needle.size();
        if (!n) {
            return 0;
        }
        vector<int> lps = kmpProcess(needle);
        for (int i = 0, j = 0; i < m;) {
            if (haystack[i] == needle[j]) { 
                i++, j++;
            }
            if (j == n) {
                return i - j;
            }
            if (i < m && haystack[i] != needle[j]) {
                j ? j = lps[j - 1] : i++;
            }
        }
        return -1;
    }
    vector<int> kmpProcess(string needle) {
        int n = needle.size();
        vector<int> lps(n, 0);
        for (int i = 1, len = 0; i < n;) {
            if (needle[i] == needle[len]) {
                lps[i++] = ++len;
            } else if (len) {
                len = lps[len - 1];
            } else {
                lps[i++] = 0;
            }
        }
        return lps;
    }
    int divide(int dividend, int divisor) {
        if (dividend == 0x80000000 && divisor == -1)return 0x7FFFFFFF;
        return dividend / divisor;
    }
};
int main(){
    printf("run success!");
    return 0;
}

File: Go/leetcode/3.go
package leetcode

type ListNode struct {
    Val int
    Next *ListNode
}

// 每相邻k个元素进行反转，而非交换元素
func reverseKGroup(head *ListNode, k int) *ListNode {
    node := head
    for i :=0; i < k; i++ {
        if node == nil {
            return head
        }
        node = node.Next
    }
    newHead := reverse(head, node)
    head.Next = reverseKGroup(node, k)
    return newHead
}
func reverse(first *ListNode, last *ListNode) *ListNode {
    prev := last
    for first != last {
        tmp := first.Next
        first.Next = prev
        prev = first
        first = tmp
    }
    return prev
}
func strStr(haystack string, needle string) int {
    // j代表needle的索引，i+j代表haystack的索引
    for i := 0; ; i++ {
        for j := 0; ; j++ {
            if j == len(needle) {
                return i
            }
            if i + j == len(haystack) {
                return -1
            }
            if needle[j] != haystack[i+j] {
                break
            }
        }
    }
}

func findSubstring(s string, words []string) []int {
    if len(words) == 0 {
        return []int{}
    }
    res := []int{}
    counter := map[string]int{}
    for _, w := range words {
        counter[w]++
    }
    length, totalLen, tmpCounter := len(words[0]), len(words[0])*len(words), copyMap(counter)
    for i, start := 0, 0; i < len(s)-length+1 && start < len(s) - length+1; i++ {
        if tmpCounter[s[i:i+length]] > 0 {
            tmpCounter[s[i:i+length]]--
            if checkWords(tmpCounter) && (i+length-start == totalLen) {
                res = append(res, start)
                continue
            }
            i = i+length-1
        }else {
            start++
            i = start - 1
            tmpCounter = copyMap(counter)
        }
    }
    return res
}

func checkWords(s map[string]int) bool {
    flag := true
    for _, v := range s {
        if v > 0 {
            flag = false
            break
        }
    }
    return flag
}
func copyMap(s map[string]int) map[string]int {
    c := map[string]int{}
    for k, v := range s {
        c[k] = v
    }
    return c
}
